src/jackdaw/test/transports/kafka.clj

```diff
@@ -0,0 +1,201 @@
+  "Clojurize the ConsumerRecord returned from consuming a kafka record"
+  [^ConsumerRecord consumer-record]
+  (when consumer-record
+    {:checksum (.checksum consumer-record)
+     :key (.key consumer-record)
+     :offset (.offset consumer-record)
+     :partition (.partition consumer-record)
+     :serializedKeySize (.serializedKeySize consumer-record)
+     :serializedValueSize (.serializedValueSize consumer-record)
+     :timestamp (.timestamp consumer-record)
+     :topic (.topic consumer-record)
+     :value (.value consumer-record)}))
+
+(defn ^ProducerRecord mk-producer-record
+  "Creates a kafka ProducerRecord for use with `send!`."
+  ([{:keys [topic-name]} value]
+   (ProducerRecord. ^String topic-name value))
+  ([{:keys [topic-name]} key value]
+   (ProducerRecord. ^String topic-name key value))
+  ([{:keys [topic-name]} partition key value]
+   (ProducerRecord. ^String topic-name ^Integer partition key value))
+  ([{:keys [topic-name]} partition timestamp key value]
+   (ProducerRecord. ^String topic-name ^Integer partition ^Long timestamp key value)))
+
+(defn consumer
+  "Creates an asynchronous Kafka Consumer of all topics defined in the
+   supplied `topic-metadata`
+
+   Puts all messages on the channel in the returned response. It is the
+   responsibility of the caller to arrange for the read the channel to
+   be read by some other process.
+
+   Must be closed with `close-consumer` when no longer required"
+  [kafka-config topic-metadata deserializers]
+  (let [continue?   (atom true)
+        messages    (async/chan 1 (comp
+                                   (map #'mk-consumer-record)
+                                   (map #(apply-deserializers deserializers %))))
+        started?    (promise)
+        poll        (poller messages)]
+
+    {:process (async/go-loop [consumer (subscription kafka-config
+                                                     (vals topic-metadata))]
+                (when-not (realized? started?)
+                  (deliver started? true)
+                  (log/infof "started kafka consumer: %s"
+                             (select-keys kafka-config ["bootstrap.servers" "group.id"])))
+
+                (if @continue?
+                  (do (poll consumer)
+                      (recur consumer))
+                  (do
+                    (async/close! messages)
+                    (.close consumer)
+                    (log/infof "stopped kafka consumer: %s"
+                               (select-keys kafka-config ["bootstrap.servers" "group.id"])))))
+     :started? started?
+     :messages messages
+     :continue? continue?}))
+
+(defn close-consumer
+  [consumer]
+  (reset! (:continue? consumer) false)
+  (async/<!! (:process consumer)))
+
+(defn build-record
+  "Builds a Kafka Producer and assoc it onto the message map"
+  [m]
+  (let [rec (mk-producer-record (:topic m)
+                                (:partition m (int 0)) ;; TODO proper partitioning ...
```

Comment:
TODO FIXME!
Currently no messages have the partition set and there is no code to try and work the partition out. Need some way to inject the correct / default partitioner into test machine writes

Reply:
It might be a good point to re-think the syntax of write commands a bit. Currently a write command is like `[:write topic-map k v timestamp timeout]`, which worked ok for our original specific use-case but I think we can improve on this to easily support a wider variety of use-cases.

e.g. in the examples below...

 record-map could include values for keys `:timestamp`, `:key`, `:value`, `:partition` etc
 write-option-map could be used to configure things like timeout, message based key/partition generation, or anything else we might want to configure in the future.

```
[:write topic-map record-map timeout]
[:write topic-map record-map write-option-map]
```

Reply:
Agreed - when giving it a go today a few of these came up immediately (more control over partitions / keys etc). Lets try to catch up tmrw.